### PR TITLE
Cache the miner state

### PR DESCRIPTION
**Pull request by elichai**
This allows a miner to cache the matrix generation to improve the performance of the miner.
I implemented a `MinerCache` that caches: the matrix, the target, the PrePowHash.

### CODE REVIEW

File: domain/consensus/utils/pow/pow.go [hunk 1]
```diff
@@ -12,52 +12,76 @@ import (
...
 
-// CheckProofOfWorkWithTarget check's if the block has a valid PoW according to the provided target
-// it does not check if the difficulty itself is valid or less than the maximum for the appropriate network
-func CheckProofOfWorkWithTarget(header externalapi.MutableBlockHeader, target *big.Int) bool {
-	// The block pow must be less than the claimed target
-	powNum := CalculateProofOfWorkValue(header)
-
-	// The block hash must be less or equal than the claimed target.
-	return powNum.Cmp(target) <= 0
+type MinerState struct {
```
Comment (someone235) [inline domain/consensus/utils/pow/pow.go:15]: Can you rename it to something more mining agnostic? Maybe just `State`? (and then it'll be `pow.State`).
It's used as part of validation, so the mining terminology feels unrelated.

File: domain/consensus/processes/blockvalidator/pruning_violation_proof_of_work_and_difficulty.go [hunk 2]
```diff
@@ -163,7 +163,7 @@ func (v *blockValidator) checkProofOfWork(header externalapi.BlockHeader) error
 
 	// The block pow must be valid unless the flag to avoid proof of work checks is set.
 	if !v.skipPoW {
-		valid := pow.CheckProofOfWorkWithTarget(header.ToMutable(), target)
+		valid := state.CheckProofOfWork()
```
Comment (someone235) [inline domain/consensus/processes/blockvalidator/pruning_violation_proof_of_work_and_difficulty.go:166]: Maybe we should still keep some wrapper like `CheckProofOfWorkWithTarget` to make it more readable? The state here isn't reused.

File: domain/consensus/utils/pow/pow.go [hunk 3]
```diff
@@ -78,7 +102,7 @@ func BlockLevel(header externalapi.BlockHeader) int {
 		return constants.MaxBlockLevel
 	}
 
-	proofOfWorkValue := CalculateProofOfWorkValue(header.ToMutable())
+	proofOfWorkValue := NewMinerState(header.ToMutable()).CalculateProofOfWorkValue()
```
Comment (someone235) [inline domain/consensus/utils/pow/pow.go:105]: nit: Use a wrapper

**Review by someone235 — CHANGES_REQUESTED** [review]
Notice that you also need to `go fmt`

File: domain/consensus/processes/blockvalidator/pruning_violation_proof_of_work_and_difficulty.go (hunk 2 above)
Comment (elichai) [reply domain/consensus/processes/blockvalidator/pruning_violation_proof_of_work_and_difficulty.go:166]: The state is used here, for the target.
I agree that a wrapper for no target is nice because it's simple, but a wrapper for no target, and a wrapper for with target, and maybe some other wrapper(for with matrix or with prePowHash) are exactly the reason I made this state :)

File: domain/consensus/utils/pow/pow.go (hunk 3 above)
Comment (elichai) [reply domain/consensus/utils/pow/pow.go:105]: That's inside of the `pow` package, do you really feel it requires a wrapper?

**Review by someone235 — APPROVED** [review]
